File: src/ast.c
 #define _GNU_SOURCE
#include <stdlib.h>
#include <unistd.h>
#include "ast.h"

void free_ast(ASTNode *node);

void free_command(Command *cmd) {
    if (!cmd) return;
    if (cmd->argv) {
        for (int i = 0; i < cmd->argc; i++)
            free(cmd->argv[i]);
        free(cmd->argv);
    }
    if (cmd->infile) free(cmd->infile);
    if (cmd->outfile) free(cmd->outfile);
    if (cmd->heredoc_delim) free(cmd->heredoc_delim);
    if (cmd->heredoc_fd != -1) close(cmd->heredoc_fd);
    free(cmd);
}

void free_pipeline(Pipeline *p) {
    if (!p) return;
    if (p->nodes) {
        for (int i = 0; i < p->count; i++)
            free_ast(p->nodes[i]);
    }
    free(p->nodes);
    free(p);
}

void free_ast(ASTNode *node) {
    if (!node) return;
    switch (node->type) {
        case NODE_COMMAND:
            free_command(node->data.cmd);
            break;
        case NODE_PIPELINE:
            free_pipeline(node->data.pipeline);
            break;
        case NODE_AND:
        case NODE_OR:
        case NODE_SEMICOLON:
            free_ast(node->data.binary.left);
            free_ast(node->data.binary.right);
            break;
        case NODE_IF:
            free_ast(node->data.if_stmt.condition);
            free_ast(node->data.if_stmt.then_body);
            free_ast(node->data.if_stmt.else_body);
            break;
        case NODE_WHILE:
            free_ast(node->data.while_stmt.condition);
            free_ast(node->data.while_stmt.body);
            break;
        case NODE_FOR:
            free(node->data.for_stmt.var);
            if (node->data.for_stmt.wordlist) {
                for (int i = 0; i < node->data.for_stmt.wordcount; i++)
                    free(node->data.for_stmt.wordlist[i]);
                free(node->data.for_stmt.wordlist);
            }
            free_ast(node->data.for_stmt.body);
            break;
        case NODE_GROUP:
            free_ast(node->data.group.body);
            break;
        default:
            break;
    }
    free(node);
}

File: src/ast.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Command {
    pub argv: Vec<String>,
    pub env_assigns: Vec<String>,
    pub infile: Option<String>,
    pub outfile: Option<String>,
    pub append_out: bool,
    pub heredoc_delim: Option<String>,
}

impl Command {
    pub fn new() -> Self {
        Command {
            argv: Vec::new(),
            env_assigns: Vec::new(),
            infile: None,
            outfile: None,
            append_out: false,
            heredoc_delim: None,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.argv.is_empty()
            && self.env_assigns.is_empty()
            && self.infile.is_none()
            && self.outfile.is_none()
            && self.heredoc_delim.is_none()
    }
}

impl Default for Command {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Pipeline {
    pub commands: Vec<Command>,
}

impl Pipeline {
    pub fn new() -> Self {
        Pipeline {
            commands: Vec::new(),
        }
    }
}

impl Default for Pipeline {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IfStmt {
    pub condition: Box<AstNode>,
    pub then_body: Box<AstNode>,
    pub else_body: Option<Box<AstNode>>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WhileStmt {
    pub condition: Box<AstNode>,
    pub body: Box<AstNode>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ForStmt {
    pub var: String,
    pub wordlist: Vec<String>,
    pub body: Box<AstNode>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AstNode {
    Command {
        command: Command,
    },
    Pipeline {
        pipeline: Pipeline,
    },
    And {
        left: Box<AstNode>,
        right: Box<AstNode>,
    },
    Or {
        left: Box<AstNode>,
        right: Box<AstNode>,
    },
    Semicolon {
        left: Box<AstNode>,
        right: Box<AstNode>,
    },
    Background {
        body: Box<AstNode>,
    },
    #[allow(dead_code)]
    If(IfStmt),
    #[allow(dead_code)]
    While(WhileStmt),
    #[allow(dead_code)]
    For(ForStmt),
    Group {
        body: Box<AstNode>,
    },
}
impl AstNode {
    pub fn command(command: Command) -> Self {
        AstNode::Command { command }
    }

    pub fn pipeline(pipeline: Pipeline) -> Self {
        AstNode::Pipeline { pipeline }
    }

    pub fn binary_and(left: AstNode, right: AstNode) -> Self {
        AstNode::And {
            left: Box::new(left),
            right: Box::new(right),
        }
    }

    pub fn binary_or(left: AstNode, right: AstNode) -> Self {
        AstNode::Or {
            left: Box::new(left),
            right: Box::new(right),
        }
    }

    pub fn binary_semicolon(left: AstNode, right: AstNode) -> Self {
        AstNode::Semicolon {
            left: Box::new(left),
            right: Box::new(right),
        }
    }

    pub fn background(body: AstNode) -> Self {
        AstNode::Background {
            body: Box::new(body),
        }
    }

    pub fn group(body: AstNode) -> Self {
        AstNode::Group {
            body: Box::new(body),
        }
    }
}
